core/src/main/java/cucumber/runtime/arquillian/ArquillianCucumber.java

```diff
@@ -350,6 +242,73 @@ public void runAfterHooks(final Reporter reporter, final Set<Tag> tags) {
         }
 
         return filters;
+    } 
+    
+    private static List<CucumberFeature> buildFeatureList(final Set<Object> testFilters, final InputStream featuresInputStream, final ClassLoader classLoader, final Set<Entry<String, Collection<URL>>> featuresSet) throws Exception
+    {
+    	final List<CucumberFeature> cucumberFeatures = new ArrayList<CucumberFeature>();
+        final FeatureBuilder featureBuilder = new FeatureBuilder(cucumberFeatures);
+        
+        if (featuresInputStream != null) {
+            final BufferedReader featuresFileReader = new BufferedReader(new InputStreamReader(featuresInputStream));
+            buildFeatureListFromFile(featuresFileReader, testFilters, featureBuilder, classLoader);            
+        }
+        else 
+        {
+        	buildFeatureListFromMap(featuresSet, testFilters, featureBuilder);
+        }
+        
+        featureBuilder.close(); 
+        
+        if (cucumberFeatures.isEmpty()) {
+            throw new IllegalArgumentException("No feature found");
+        }
+        
+        return cucumberFeatures;
+    }
+    
+    private static void buildFeatureListFromFile(final BufferedReader featuresFileReader, final Set<Object> testFilters, final FeatureBuilder featureBuilder, final ClassLoader classLoader) throws Exception
+    {
+    	String readerLine;
+
+        while ((readerLine = featuresFileReader.readLine()) != null) {
+            readerLine = readerLine.trim();
+            if (readerLine.isEmpty()) {
+                continue;
+            }
+
+            final PathWithLines pathWithLines = new PathWithLines(readerLine);
+            testFilters.addAll(pathWithLines.lines);
+            featureBuilder.parse(new ClassLoaderResource(classLoader, pathWithLines.path), new ArrayList<Object>(testFilters));
+        }
+    }
+    
+    private static void buildFeatureListFromMap(final Set<Entry<String, Collection<URL>>> featuresSet, final Set<Object> testFilters,  final FeatureBuilder featureBuilder)
```

Comment:
a Set<Entry> is not common where Map is and easy to debug, I'd use Map and map.entrySet() to iterate over it, wdyt?